README.md

```diff
@@ -0,0 +1,84 @@
+
+## Motivation
+[ARCX](https://github.com/algorandfoundation/ARCs/pull/229) provides a standard for delegated asset opt-ins, but there are some UX problems that need to be addressed. First, there needs to be a way to control how long the signature is valid for. Without an application to control this, signing the logic signature program would be irreversible. There also needs to be a standardized way to store and read signatures for a given account so dApps and other users can take advantaged of delegated opt-ins.
+
+## Specification
+This is an implementation of [ARCX](https://github.com/algorandfoundation/ARCs/pull/229). There is additional functionality provided by the application in this ARC in the methods of the application.
+
+For all methods, refer to the [ABI JSON description](./contracts/artifacts/DelegatedOptIn.abi.json) for complete descriptions of arguments.
+
+### Signature Storage
+
+`setOpenOptInSignature(byte[64],address,txn)void` and `setAddressOptInSignature(byte[64],address,address,txn)void` are methods for adding signatures to box storage for open opt-ins and address opt-ins, respectively.
+
+Both methods utilize [the verifier lsig](./contracts/verifier_lsig.teal) to verify the address being added to box storage matches the address(es) used as the key. This means anyone can query the applications boxes to get the signature for the given address(es) and be certain that it is correct.
+
+### End Times
+
+`setOpenOptInEndTime(uint64)void` and `setAddressOptInEndTime(uint64,address)void` are methods for setting the end time of a signature for open opt-ins and address opt-ins, respectively. An endtime signifies when the delegated logic signature will no longer work. The time corresponds to the epoch time (seconds) returned by `global LatestTimestamp`. End times can be updated at any time to any value.
+
+### Opt-Ins
+
+`openOptIn(pay,axfer)void` and `addressOptIn(pay,axfer)void` are implementations of the [ARCX](https://github.com/algorandfoundation/ARCs/pull/229) interfaces. They both verify the MBR payment is sent to the account opting in and that it covers the ASA minimum balance requirement, which is stored in global storage. It also verifies the value of `global LatestTimestamp` is less than the set end time for the account opting in.
+
+### Storage
+
+| Type | Key | Description |
+| ---- | --- | ----------- |
+| Global | "sigVerificationAddress" | Stores the address of [the verifier lsig](./contracts/verifier_lsig.teal) |
+| Global | "assetMBR" | Stores the ASA MBR |
+| Box | "s-" + `auth-addr` | Mapping of signer to open opt-in signature |
+| Box |  "e-" + `auth-addr` | Mapping of signer to open opt-in end time |
+| Box | "s-" + `sha256(signer,sender)` | Mapping of the hash of the signer address plus sender address to address opt-in signature |
```

Comment:
The hash is used because there are two key-value pairs that use the `(sender,signer)` tuple as a key (end times and signature), thus a prefix is needed to differentiate them which results in more than 64 bytes. 

There could be a single box that stores both the signature and the end time under a single key, but that does add some extra complexity to decoding that data structure (vs just checking box existence).

Reply:
If the signature box got to be the one named by the concatenation, then you'd enable the kind of lookup I'm describing.

Reply:
Ah true didn't think of that. Will update so the key for signatures is just the non-hashed `(sender,signer)` tuple